[Review] : 리뷰 수정 기능 구현

rlqor9688 opened the pull request:
## Description
1. 리뷰 수정 기능을 구현하였습니다.
2. 리뷰 수정은 PatchMapping입니다. (일부 필드만 수정하기 때문)
3. 리뷰 수정 시 수정할 reviewId, authUser, ReviewUpdateRequestDto를 입력으로 받아옵니다.

## Changes
- 1. ReviewController 내 updateReview 메서드 구현
- 2.  ReviewService 인터페이스 내 updateReview 메서드 추가
- 3. ReviewServiceImpl 클래스 내 updateReview 메서드 구현
- 4. ReviewServiceUpdateTest 내 테스트 코드 구현
- 5. ReviewUpdateRequestDto 클래스 신규 생성
- 6. Reviews Entity 수정
- 7. 기존 ReviewServiceImplTest를 ReviewServiceCreateTest로 클래스명 변경
- 8. ErrorCode 추가

## Ref
### 테스트 결과
![image](https://github.com/user-attachments/assets/4e4e1d51-41f7-4e1d-8753-4143ff44fb88)

rlqor9688 committed:
feat [review] : 리뷰 수정 기능 구현

1. ReviewController 내 updateReview 메서드 구현

2. ReviewServiceImpl 클래스 내 updateReview 메서드 구현

3. ReviewServiceUpdateTest 내 테스트 코드 구현

4. 기존 ReviewServiceImplTest를 ReviewServiceCreateTest로 변경

rlqor9688 committed:
docs [review] : 리뷰 수정 기능 RequestMapping 메서드 변경

1. API설계 상 PutMapping이나, PatchMapping으로 변경 (리소스 일부 필드만 변경하기 때문)

nagul2 commented on src/main/java/com/delivery/igo/igo_delivery/api/review/controller/ReviewController.java:
```diff
@@ -27,4 +26,14 @@ public ResponseEntity<ReviewResponseDto> createReview(
+            @PathVariable Long reviewId,
+            @Auth AuthUser authUser,
+            ReviewUpdateRequestDto requestDto
+    ) {
+        reviewService.updateReview(reviewId, authUser, requestDto);
+        return new ResponseEntity<>(HttpStatus.OK);
+    }
```
204 응답으로 수정해서 보내는 것이 좋을 것 같습니다. 명세서는 200인데 수정이 필요한 것 같아요

rlqor9688 replied:
감사합니다! 반영 완료

nagul2 commented on src/main/java/com/delivery/igo/igo_delivery/api/review/entity/Reviews.java:
```diff
@@ -57,6 +57,11 @@ public void delete() {
         this.reviewStatus = ReviewStatus.DELETED;
     }
 
+    public void update(String content, Integer rating) {
+        this.content = content;
+        this.rating = rating;
+    }
```
까서 넘기셨군요..ㅋㅋㅋㅋ

rlqor9688 replied:
ㅋㅋㅋㅋ 좋다는 건 일단 써봐야.. 100% 이해하진 못했지만 dto 의존성을 줄인다는 논리에 공감했습니다!

nagul2 commented on src/main/java/com/delivery/igo/igo_delivery/api/review/service/ReviewServiceImpl.java:
```diff
@@ -90,28 +90,39 @@ public ReviewResponseDto createReview(AuthUser authUser, ReviewRequestDto reques
+    }
+
+    // 유저 유효성 검증(UserStatus=LIVE, UserRole = CONSUMER)
+    private void validateUser(Users user) {
+        user.validateDelete();
+        user.validateConsumer();
+    }
```
private 메서드는 더 명확하게 행위를 나타내는게 좋다고 합니다.
저도 그래서 수정했기 때문에 role, 삭제를 검증한다는 검증행위가 추가로 들어가면 좋을 것 같습니다.

rlqor9688 replied:
감사합니다! 피드백 반영 완료!

anyubinn commented on src/main/java/com/delivery/igo/igo_delivery/api/review/service/ReviewServiceImpl.java:
```diff
@@ -90,28 +90,39 @@ public ReviewResponseDto createReview(AuthUser authUser, ReviewRequestDto reques
+        validateReviewAccess(findReview, authUser);
+
+        // 로그인 유저 DB 조회 + 상태(LIVE), 권한(CONSUMER) 확인
+        Users findUser = userRepository.findById(authUser.getId())
```
현재는 리뷰를 먼저 조회한 뒤 유저를 검증하고 있는데 로그인한 유저의 상태(LIVE, CONSUMER)를 먼저 검증한 뒤 조건을 만족하는 경우에만 리뷰에 접근하도록 하면 더 안전하고 명확한 흐름이 될 것 같아요!

rlqor9688 replied:
감사합니다! 반영 완료!

rlqor9688 committed:
Merge branch 'dev' into feature/kiback-review

rlqor9688 committed:
docs [review] : 리뷰 수정 기능 응답코드, 로직 구현 순서 수정

1. ReviewController : 성공 응답 코드 200에서 204로 수정

2. ReviewServiceImpl : Private 메서드명이 기능을 포함하도록 구체화 수정, 유효성 검증 로직 수정

3. ReviewServiceUpdateTest : 유효성 검증 로직 수정에 따른 given 입력 데이터 수정

nagul2 commented:
제 PR이 먼저 승인되서 AuthUser부분 충돌 나시는 것같으니 dev 머지해서 충돌 해결 보시면 될 것 같아요

rlqor9688 committed:
Merge branch 'dev' into feature/kiback-review

rlqor9688 committed:
docs [review] : 최신 Pull 이후 Merge

1. ReviewServiceImpl : Users의 validateAccess 필드값 변경에 따른 서비스 메서드 수정
